// xr-utils/lib/fetch/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.get = get;
exports.post = post;
exports.test = test;

require('whatwg-fetch');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function get(url, headers) {
	headers = (0, _extends3.default)({
		method: 'GET',
		headers: {
			'Accept': 'application/json',
			'Content-Type': 'application/json'
		}
	}, headers);

	return new _promise2.default(function (resolve, reject) {
		fetch(url, headers).then(function (response) {
			return response.json();
		}).then(function (responseJson) {
			return resolve(responseJson);
		}).catch(function (error) {
			return reject(error);
		});
	});
}

function post(url, data, headers) {
	headers = (0, _extends3.default)({
		method: 'POST',
		headers: {
			'Accept': 'application/json',
			'Content-Type': 'application/json'
		}
	}, headers, {
		body: (0, _stringify2.default)(data)
	});

	return new _promise2.default(function (resolve, reject) {
		fetch(url, headers).then(function (response) {
			return response.json();
		}).then(function (responseJson) {
			return resolve(responseJson);
		}).catch(function (error) {
			return reject(error);
		});
	});
}

function test(url, data, result) {
	return new _promise2.default(function (resolve, reject) {
		setTimeout(function () {
			resolve(result);
		}, 0);
	});
}

exports.default = {
	fetch: fetch,
	get: get,
	post: post,
	test: test
};
